tensorboard/data/proto/data_provider.proto

```diff
@@ -0,0 +1,295 @@
+syntax = "proto3";
+
+package tensorboard.data;
+
+import "tensorboard/compat/proto/summary.proto";
+import "tensorboard/compat/proto/tensor.proto";
+
+option go_package = "github.com/tensorflow/tensorboard/tensorboard/data/proto/data_provider_proto";
+
+service TensorBoardDataProvider {
+  // List runs within an experiment.
+  rpc ListRuns(ListRunsRequest) returns (ListRunsResponse) {}
+  // List metadata about scalar time series.
+  rpc ListScalars(ListScalarsRequest) returns (ListScalarsResponse) {}
+  // Read data from scalar time series.
+  rpc ReadScalars(ReadScalarsRequest) returns (ReadScalarsResponse) {}
+  // List metadata about tensor time series.
+  rpc ListTensors(ListTensorsRequest) returns (ListTensorsResponse) {}
+  // Read data from tensor time series.
+  rpc ReadTensors(ReadTensorsRequest) returns (ReadTensorsResponse) {}
+  // List metadata about blob sequence time series.
+  rpc ListBlobSequences(ListBlobSequencesRequest)
+      returns (ListBlobSequencesResponse) {}
+  // Read blob references from blob sequence time series. See `ReadBlob` to read
+  // the actual blob data.
+  rpc ReadBlobSequences(ReadBlobSequencesRequest)
+      returns (ReadBlobSequencesResponse) {}
+  // Read data for a specific blob.
+  rpc ReadBlob(ReadBlobRequest) returns (stream ReadBlobResponse) {}
+}
+
+message ListRunsRequest {
+  // ID of experiment in which to query data.
+  string experiment_id = 1;
+}
+
+message ListRunsResponse {
+  repeated Run runs = 1;
+}
+
+message Run {
+  // Unique opaque string identifier.
+  string id = 1;
+  // User-facing name.
+  string name = 2;
+  // Wall time of earliest recorded event, as floating-point seconds since
+  // epoch (same as event file format).
+  double start_time = 3;
+}
+
+message ListScalarsRequest {
+  // ID of experiment in which to query data.
+  string experiment_id = 1;
+  // Required filter for plugin name. If omitted, an empty message is implied.
```

Comment:
Agh… I’m so torn. I agree, and have been burned by this myself when
using `grpc_cli` to send requests (to TensorBoard.dev, which has the
same semantics) without specifying `downsample` and wondering why the
response was empty. But I’ve also been burned so many times by systems
that treat zero as not a number when there’s clearly a continuous
interpretation of the specification.

I’m happy to make `plugin_name` an error, since requiring that plugin
names not be empty seems perfectly reasonable. We already require that
_dashboard_ names not be empty, which is related, though distinct. But
empty and non-empty strings differ in relevant ways (e.g., disappearing
from path components, not making inductive progress) that zero and
positive integers don’t. And it’s so easy to imagine user flows that
_do_ want to set `downsample { num_points: 0 }`:

> **TensorBoard Export-Your-Data™ Wizard**
>
> How many points do you want from each of these time series?
>
>   - *accuracy*: `1000`
>   - *gan_output/image/0*: `10`
>   - *layer0/weights*: `0` &#x2003;**💥 kaboom!**
>
> <kbd>Download Archive</kbd>

I’ll break the tie by deferring to your intuition and making both cases
an error, but I would be happy to change `downsample` back to continuous
if we change our minds.

Reply:
Actually I wasn't suggesting that we make `downsample { num_points: 0 }` an error; I was suggesting only that we make the absence of `downsample` entirely an error (sorry if this was ambiguous).  So code that populates the `Downsample` message has no discontinuity at zero; it's only code that doesn't set `Downsample` entirely.  That takes care of the `grpc_cli` case.

Reply:
Ah, got it. Yeah, that sounds pretty reasonable. I generally like the
semantics of “omitted messages are implied to be empty”, but I’m happy
to deviate from it where required or sensible (e.g., `RunTagFilter`),
and this seems like such an occasion. Sent you #4333.